### PR TITLE
Fix kafka performance test write failing due to server overwhelmed

### PR DESCRIPTION
**Pull request by Abacn**
Part of #22303

Kafka performance test starts failing again after #24879 in. This is because we removed shuffle=appliance there and then read from shuffle becomes faster. However, the producer is unable to digests data within timeout (see flooding warning log of `send failed : 'Expiring 148 record(s) for beam-sdf-0:130675 ms has passed since batch creation'`) 

entered #24963 for the IO connector fix. For now remove ReShuffle and make the test green again.

* Remove ReShuffle before write

* Minor Optimizations on WriteCallBack

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Abacn** [comment]
R: @johnjcasey

**Comment by Abacn** [comment]
Run Java KafkaIO Performance Test

**Comment by Abacn** [comment]
Performance test succeeded after removing Reshuffle (on test branch)

https://ci-beam.apache.org/job/beam_PerformanceTests_Kafka_IO/3538/


Seems like jenkins trigger broken again

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriter.java [hunk 1]
```diff
@@ -42,13 +42,16 @@
  */
 class KafkaWriter<K, V> extends DoFn<ProducerRecord<K, V>, Void> {
 
+  protected transient @Nullable Callback callback;
```
Comment (johnjcasey) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriter.java:45]: what is the motivation for a single callback?

**Comment by Abacn** [comment]
Run RAT PreCommit

**Comment by Abacn** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Comment by Abacn** [comment]
Run Java KafkaIO Performance Test

**Comment by Abacn** [comment]
Run Python_PVR_Flink PreCommit

**Comment by Abacn** [comment]
hold on for now. Removing the reshuffle in testbatch write causing batch test has high probability fail (reshuffle added before to fix this issue)

We need to fix the IO connector to make sure it is robust to incoming traffic

**Comment by Abacn** [comment]
Run Java KafkaIO Performance Test

**Comment by Abacn** [comment]
Run Java KafkaIO Performance Test

**Comment by Abacn** [comment]
test succeeded. Anyway flaky is better than permared. PTAL @johnjcasey

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 2]
```diff
@@ -212,14 +210,16 @@ public void testKafkaIOReadsAndWritesCorrectlyInStreaming() throws IOException {
 
     PipelineResult writeResult = writePipeline.run();
     PipelineResult.State writeState = writeResult.waitUntilFinish();
+    // Fail the test if pipeline failed.
+    assertNotEquals(PipelineResult.State.FAILED, writeState);
```
Comment (Abacn) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java:214]: Current failing on master is due to the write pipeline continuously hitting OOM and cannot finish the bundle within timeout; then read pipeline does not receive expected number of data, e.g. https://ci-beam.apache.org/view/PerformanceTests/job/beam_PerformanceTests_Kafka_IO/3546/console
```
org.apache.beam.sdk.io.kafka.KafkaIOIT > testKafkaIOReadsAndWritesCorrectlyInStreaming FAILED
20:26:04     java.lang.AssertionError: actual number of records 34569527 smaller than expected: 100000000.
```
put this assertion right after writeResult.waitUntilFinish() to make test fail early if write pipeline fails.

The OOM issue is also a motivation of using single Callback instance (though not have decisive help)

**Review by johnjcasey — APPROVED** [review]

**Comment by Abacn** [comment]
GA Playground Examples CI failure irrelevant. merging for now

**Pull request merged**